### PR TITLE
types(node): add changes from 10.10

### PR DESCRIPTION
**Pull request by SimonSchick**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/nodejs/node/blob/master/doc/changelogs/CHANGELOG_V10.md#2018-09-06-version-10100-current-targos
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

RE: #28708

Lets try this again.

### CODE REVIEW

**Comment by SimonSchick** [comment]
@andy-ms this PR and the related work was extremely painful, it would be great if we some core contributor could bump up the minimum TS version to something like 2.9

**Comment by ZaneHannanAU** [comment]
Indeed.

**Review by ZaneHannanAU — APPROVED** [review]

File: types/node/index.d.ts [hunk 1]
```diff
@@ -795,6 +795,14 @@ declare namespace NodeJS {
         disconnect(): void;
         connected: boolean;
 
+        /**
+         * The `process.allowedNodeEnvironmentFlags` property is a special,
+         * read-only `Set` of flags allowable within the [`NODE_OPTIONS`][]
+         * environment variable.
+         */
+        // TODO: This is readonly
+        allowedNodeEnvironmentFlags: Set<string>;
```
Comment (r3nya) [inline types/node/index.d.ts]: > TODO

Maybe we can use this way?

`readonly allowedNodeEnvironmentFlags: Set<string>;`

**Review by r3nya — APPROVED** [review]
Great job!

File: types/node/index.d.ts (hunk 1 above)
Comment (SimonSchick) [reply types/node/index.d.ts]: Eh, yea this technically readonly, the comment is about the actual `Set` tho.

**Comment by SimonSchick** [comment]
Travis failed for some reason, do I need to push to re-run the build?

File: types/node/index.d.ts (hunk 1 above)
Comment (ZaneHannanAU) [reply types/node/index.d.ts]: We could throw in `ReadonlySet`, `ReadonlyMap` and so on.

Basically just a `get` and `has` without the `set` and `add` functions of normal `Map`/`WeakMap` and `Set`/`WeakSet`s.

**Comment by ZaneHannanAU** [comment]
```shell
curl --http2 -s --compressed https://api.travis-ci.org/v3/job/429175505/log.txt | grep -i error -C 4
```

cool no real errors now

File: types/node/index.d.ts (hunk 1 above)
Comment (SimonSchick) [reply types/node/index.d.ts]: I don’t think we should duplicate tbese inbuilt type definitions, probably better to wait for node typings to be bumbed up to a higher ts version

File: types/node/index.d.ts [hunk 2]
```diff
@@ -4379,6 +4467,22 @@ declare module "fs" {
...
+     * URL support is _experimental_.
+     * If a file descriptor is provided, the underlying file will _not_ be closed automatically.
+     * @param data The data to write.
+     * @param options Either the encoding for the file, or an object optionally specifying the encoding, file mode, and flag.
+     * If `encoding` is not supplied, the default of `'utf8'` is used.
+     * If `mode` is not supplied, the default of `0o666` is used.
+     * If `mode` is a string, it is parsed as an octal integer.
+     * If `flag` is not supplied, the default of `'w'` is used.
+     */
+    export function writeFile(path: PathLike | number, data: BinaryData, options: WriteFileOptions, callback: (err: NodeJS.ErrnoException) => void): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: I still think we don't need this overload here as the other below has `any` and the documentation there states already `If something other than a Buffer or Uint8Array is provided, the value is coerced to a string.`

File: types/node/index.d.ts [hunk 3]
```diff
@@ -4431,7 +4535,7 @@ declare module "fs" {
      * If `mode` is a string, it is parsed as an octal integer.
      * If `flag` is not supplied, the default of `'w'` is used.
      */
-    export function writeFileSync(path: PathLike | number, data: any, options?: { encoding?: string | null; mode?: number | string; flag?: string; } | string | null): void;
+    export function writeFileSync(path: PathLike | number, data: BinaryData | string | any, options?: WriteFileOptions): void;
```
Comment (Flarna) [inline types/node/index.d.ts]: I don't think the datatype of `data` should be changed here. As `any` is allowed adding others doesn't change anything and doc provides the needed info already.

**Review by Flarna — CHANGES_REQUESTED** [review]
Could you please add https://nodejs.org/api/os.html#os_priority_constants

**Review by Flarna — APPROVED** [review]

**Comment by RyanCavanaugh** [comment]
FYI we're working on features for TS 3.1 that will allow us to have forward-compatible versioning, so we can have a 2.6 Node package on DT while still having a TS3.1+ version

**Comment by RyanCavanaugh** [comment]
FYI we're working on features for TS 3.1 that will allow us to have forward-compatible versioning, so we can have a 2.6 Node package on DT while still having a TS3.1+ version

**Pull request merged**